### ISSUE TITLE
Idiom for Tensor subclasses that are also Parameters

### ISSUE THREAD

**Issue by ezyang** [pytorch/pytorch#65018]
Parameter is a subclass of Tensor. You might also define your own subclass of Tensor. How to also make it a Parameter? Do you have to do a multiple inheritance thing? Are there other ways to do it?

cc @albanD @mruberry @jbschlosser @walterddr @hameerabbasi @rgommers @peterbell10

**Comment by albanD** [comment]
A "safe" proposal could be:
- Make `nn.Parameter()` raise an error if you pass a subclass, directing you to use `nn.WrappingParameter`.
- Add `nn.WrappingParameter` that is a pure python implementation of a Wrapped Tensor style subclass that can contain any user-provided subclass and inherits from `nn.Parameter`.

**Comment by ezyang** [comment]
Do we want to have people do mixin style? It's being done this way in the codebase for lazy modules.

**Comment by albanD** [comment]
After some discussion with @jbschlosser what about:
```python

class Parameter:
    def __new__(cls, data):
        if cls is not Parameter:
            assert super() is not object
            super().__new__(data)
        else:
            user_class = data.__class__
            user_param_class = type(f"{user_class.__name__}Parameter", (Parameter, user_class), {"_base_class": user_class})
            return user_param_class(data)

    @classmethod
    def __torch_function__(cls, func, types, args=(), kwargs=None):
        res = super().__torch_function__(...)

        out = []
        for el in res:
            if isinstance(res, cls):
                el.__class__ = el._base_class
            out.append(el)

        return out
```

And there is an obvious fast-path for TensorParameter that we can add to make sure we don't regress perf.

**Comment by albanD** [comment]
From some offline discussion with @ezyang the above example is interesting in spirit but creating new class for every call is going to be very surprising for the user.

As a short term fix, the solution could be to raise an error when the user tries to create a Parameter wrapping a subclass, pointing to a doc page explaining how to create a new subclass of their class and Parameter.

**Comment by jbschlosser** [comment]
Dynamically generating a Parameter subclass per custom tensor type has at least one problem: such classes cannot be pickled because they are not defined at the top level. See [here](https://docs.python.org/3/library/pickle.html#what-can-be-pickled-and-unpickled) for restrictions on what can be pickled and [here](https://stackoverflow.com/a/15247892) for a terrible hacky way to inject the generated class into `globals()` (note that we'd somehow have to do such a `globals()` injection before loading a serialized parameter subclass as well).

**Comment by jbschlosser** [comment]
For posterity, I wanted to post some rationale that caused us to move away from the initial direction proposed in #62787 (having Parameter be a wrapping tensor) and towards the design discussed above:
* For vanilla tensors, the extra overhead involved with Parameter being a wrapping tensor is unacceptable. Potentially, this could be special-cased to avoid the overhead.
* For custom tensors, having Parameter be a wrapping tensor makes it difficult for it to act like the custom tensor in question. The ideal scenario is to have the capability for a custom tensor that acts exactly like the custom tensor with the additional Parameter "tag". While this is possible on paper with Parameter as a wrapping tensor, it's tricky in the details, and subclassing the custom tensor makes this much easier.

**Comment by ezyang** [comment]
Here's my take: we should come up with some member variable on Tensor such that, if it is true, the class counts as a parameter. Parameter subclass gains this member variable and we modify all of PyTorch internals to test for this member variable rather than do a subclass check. It is simple, backward compatible, and doesn't require mix-in shenanigans (and independently defining a parameter concept).

**Comment by albanD** [comment]
We still to maintain that `isinstance(foo, nn.Parameter)` work for user code no? We were working with the assumption that we do.

Also how surprising do you think it would be for the user for two instance of the same class, for one to be a Parameter and not the other?

**Comment by jbschlosser** [comment]
It's theoretically possible to implement `Parameter.__instancecheck__` in such a way that a tensor instance with such a member variable set still returns True for `isinstance(foo, Parameter)`.

> Also how surprising do you think it would be for the user for two instance of the same class, for one to be a Parameter and not the other?

Not clear to me - is there visible behavior (beyond the isinstance checks, covered above) that is likely to change for users that assume different class types between the two?

**Comment by ezyang** [comment]
Yes, I would have overridden the magic method. As for being per instance or per class, it seems nice not to have to explicitly define parameter subclass for every custom subclass. So I would vote for per instance.

**Comment by jbschlosser** [comment]
This is a minor point, but I want to bring it up regardless: if we go the route of embedding a flag to indicate parameter-ness on a tensor subclass instance, the responsibility of propagating this flag must be taken on by any custom `__deepcopy__` / `__reduce_ex__` implementations.

#73078 expands the base implementations of these magic methods to better handle wrapper subclasses and avoid the need for custom implementations in most cases, but I imagine there are cases where this can't be avoided, and if the parameter-ness flag is not correctly propagated, it will be confusing for users.

**Comment by ezyang** [comment]
Sure. This doesn't seem too bad for an end user to figure out.

**Comment by jbschlosser** [comment]
@ezyang Sounds good, but what about `__repr__`? This function should generally be implemented for each tensor subclass, and I can't think of a nice way to get what we want (`Parameter({sub_tensor.__repr__()})`) without pushing that responsibility onto the subclass writer. This seems gross to me:

```python
class LoggingTensor(torch.Tensor):
    ...
    def __repr__(self):
        # This function has to be aware of the parameter-ness flag to be implemented properly.
        rep = f"LoggingTensor({self.elem})"
        if hasattr(self, '_is_param'):
            rep = f"Parameter({rep})"
        return rep
    ...
```

**Comment by ezyang** [comment]
Good question. I think my preferred resolution is to have the built-in repr for Tensor know how to deal with printing parameter-ness. If someone overrides repr, they'll have to remember how to print parameter-ness, but that's true today anyway.

BTW, the current parameter printing is kind of goofy, it isn't valid repr:

```
>>> torch.nn.Parameter(torch.empty(2,2))
Parameter containing:
tensor([[ 0.0000e+00,  2.5244e-29],
        [-2.7151e+10,  3.6902e+19]], requires_grad=True)
```

**Comment by jbschlosser** [comment]
> Good question. I think my preferred resolution is to have the built-in repr for Tensor know how to deal with printing parameter-ness. If someone overrides repr, they'll have to remember how to print parameter-ness, but that's true today anyway.

So this is similar to what is done for serialization and deepcopying - provide a default base impl that deals with parameter-ness and is usable for the majority of subclass cases.

What does this look like in practice? For a given subclass that needs a custom repr:
* Define a `SubTensor.__repr__()`that calls `super().__repr__()` and string-manipulates the result? Seems messy and not flexible enough for cases like `SparseTensor`, which maintains both a `values` & `indices` component.
* Don't define a `SubTensor.__repr__()` and fallback to the base impl? For customization, introduce some new function (along the lines of `Module.extra_repr()`) returning a string or dict of things to print. Define this instead for `SubTensor` and use the result in the base repr impl.
* Something else?

Note that existing tensor subclasses that override `__repr__()` will have to change and won't get the parameter-awareness for free, but that's okay.

> BTW, the current parameter printing is kind of goofy, it isn't valid repr:

Yep sure enough, this is awful :) I'm planning to make it a valid repr in a (probably painful) follow-up PR but maintain the current behavior for plain tensors initially.

**Comment by ezyang** [comment]
`extra_repr` (though I wouldn't call it that) is a good idea, especially for subclasses that aren't in the "wrapping" style, because it's hard to print the tensor in a good way anyway.

**Comment by jbschlosser** [comment]
Closing as addressed by #73459.
